### DISCUSSION TITLE
ESP32-C3-USB serial only works when serial is open at hard reset

### DISCUSSION THREAD

**Discussion by TRadigk** [micropython/micropython#12267] [ESP32]
Hi everyone,

I can't say for sure the title is correct, but my findings indicate a bug specific to the USB variant of the ESP32 port. When trying to connect to the board (in this instance a https://wiki.seeedstudio.com/XIAO_ESP32C3_Getting_Started/) it is important to reset the board via the onboard reset switch and very quickly connect, e.g. via Thonny.
Taking too much time in this process will always lead to WriteTimeout on serial connection, like so:
```
PROBLEM IN THONNY'S BACK-END: Internal error (serial.serialutil.SerialTimeoutException: Write timeout).
See Thonny's backend.log for more info.
You may need to press "Stop/Restart" or hard-reset your MicroPython device and try again.


Process ended with exit code 0.
```

I also tried other programs like HTerm to discover that a connection with 115200 Baud is _not supported_ (though I think this is irrelevant?!?):
![grafik](https://github.com/micropython/micropython/assets/92845708/851a0bc3-d9ff-407a-b47e-757a756d9a5e)
And also the number of StopBits has to be configured to "1.5".

Even when the connection is possible, HTerm is not able to receive data while the board is sending data periodically. Only when performing hard reset and quickly disconnecting/connecting to the board again allows to send/receive data:
![grafik](https://github.com/micropython/micropython/assets/92845708/ae7a0ce8-664e-4cd3-bef7-d6843d6c5b57)

Why I think this an issue with the micropython port?

- Having the board in this state remains until power down / hard disconnecting it from power. Meaning ever following action now does not lead to serial write timeout.
- Using the same micropython firmware on a ESP32 with USB bridge (CH340) and same python files does not show any issue (can't repro with this board: https://docs.espressif.com/projects/esp-idf/en/latest/esp32c3/hw-reference/esp32c3/user-guide-devkitm-1.html)
- Also can't repro this issue with RP2040, though I know this is a different port, but interestingly DTR on serial has to be enabled for RP2040, while for ESP32 it must not

It "feels" like the device must see an open connection when the bootloader hands over control to micropython.

Is there anything else I should be able to provide when converting this discussion to an issue?

**Comment by karfas** [comment]
Most likely, thonny is not the right tool to test for "errors" like this, unless you know **exactly** what thonny does/tries to do with the serial connection. 
An ESP32-C3 ("GENERIC_C3_USB") doesn't have this problem with mpremote:
```
app (olimex)]$ mpremote a2 
Connected to MicroPython at /dev/ttyACM2
Use Ctrl-] to exit this shell
TX Msg #18
TX Msg #19
Traceback (most recent call last):
  File "main.py", line 63, in <module>
KeyboardInterrupt: 
MicroPython v1.20.0-349-g038f39234 on 2023-08-19; ESP32C3 module with ESP32C3
Type "help()" for more information.
>>>

**Reply by TRadigk** [reply]
I am not familiar with mpremote. 

~~But you are right. The connection is much more reliable. I will look into that.~~

When I connect with mpremote I can see my print statements coming and hitting CTRL+C during this loop will end up having mpremote getting stuck, not reading from serial anymore.

**Reply by TRadigk** [reply]
With mpremote I can easily also enter the same state:
```
Connected to MicroPython at COM3
Use Ctrl-] or Ctrl-x to exit this shell
Traceback (most recent call last):
  File "<frozen runpy>", line 198, in _run_module_as_main
  File "<frozen runpy>", line 88, in _run_code
  File "C:\Users\user\AppData\Roaming\Python\Python311\Scripts\mpremote.exe\__main__.py", line 7, in <module>
  File "C:\Users\user\AppData\Roaming\Python\Python311\site-packages\mpremote\main.py", line 510, in main
    do_repl(state, argparse_repl().parse_args([]))
  File "C:\Users\user\AppData\Roaming\Python\Python311\site-packages\mpremote\repl.py", line 79, in do_repl
    do_repl_main_loop(
  File "C:\Users\user\AppData\Roaming\Python\Python311\site-packages\mpremote\repl.py", line 30, in do_repl_main_loop
    state.pyb.serial.write(c)
  File "C:\Users\user\AppData\Local\Programs\Python\Python311\Lib\site-packages\serial\serialwin32.py", line 325, in write
    raise SerialTimeoutException('Write timeout')
serial.serialutil.SerialTimeoutException: Write timeout
```

**Comment by jimmo** [comment]
@TRadigk is this a new problem (i.e. was the board previously working), or is it the first time using the board?

I ask because we used to have two different configurations for this board which we just merged into one at the end of last week. (See #12270). You're using the new merged configuration.

**Comment by TRadigk** [comment]
Hi @jimmo 
I've encountered this issue already before the release of 1.20. I think it is related the cause of Pymaker in VS Code being unable to upload files and getting stuck while doing so. I thought I could mediate the problem by implementing a C# version of Pyboard, which now shows me that uploading might work, but I am having a hard time connecting.

The esptool encounters the same issue:
```
PS C:\Users\user\source\external\esptool> python esptool.py --chip auto --port COM3 erase_flash
esptool.py v4.6-dev
Serial port COM3
Connecting...
Detecting chip type... ESP32-C3
Chip is ESP32-C3 (revision v0.4)
Features: WiFi, BLE
Crystal is 40MHz
MAC: 34:85:18:03:17:74
Uploading stub...
Running stub...
Stub running...

A fatal error occurred: Unable to verify flash chip connection (No serial data received.).
```

(chip auto or esp32 doesn't make any difference, since the esp32 is detected correctly anyway)

**Reply by robert-hh** [reply]
Pymakr is an unsupported product of the company Pycom, which ceased operation about a year ago. Even before, Pymakr had the habit of failing with almost every  new version of VS Code. So better change horses.

**Reply by TRadigk** [reply]
Hi robert,

I liked the integration in vs code, which I use for my micropython projects. I didn't yet find any suitable replacement, but I admit I didn't look very far. But with this I found an opportunity for a new project.

@all:
~~Looking into the code of mpremote and comparing it to pyboard, which I used as a template for my C# implementation I couldn't find any difference when the port is opened. But since I am testing with boards that have some LEDs wired up I noticed that whenever the console is exited, the board does a soft_restart.
In the early stages of getting into the communication layer via stdin/stdout I noticed that triggering **Rts** on serial can have the same effect. I've now implemented this trigger on my "connect" method and _tadaa_ the connection now seems more reliable.~~

**Comment by TRadigk** [comment]
One thing still remains curious: why does **this** (USB-C based ESP32-C3) board remain _special_ in this regard? I can't get past that 🤯

Edit: The issue is very deceiving. At some point one could think "now I got it", but this is only true because I have entered the state where the connection always works. It remains: when power was disconnected, getting a stable connection is very difficult.

Edit2: The most reliable solution currently is to set RTS on, wait 300 ms and set RTS off before sending first bytes.
From my quickly improvised logging:
```
22.08.2023 11:26:18.303 [Info] [PyboardSharpTransport] : Opened connection to COM3 with Baud115K
22.08.2023 11:26:18.308 [Debug] [PyboardSharpTransport] : RTS on
22.08.2023 11:26:18.612 [Debug] [PyboardSharpTransport] : RTS off
Set Dtr to True
22.08.2023 11:26:18.689 [Debug] [PyboardSharpTransport] : sending bytes: 0x0D;0x03;0x03 ...
22.08.2023 11:26:18.690 [Debug] [PyboardSharpTransport] : sent bytes: 0x0D;0x03;0x03
22.08.2023 11:26:18.705 [Debug] [PyboardSharpTransport] : sending bytes: 0x02 ...
22.08.2023 11:26:18.705 [Debug] [PyboardSharpTransport] : sent bytes: 0x02
22.08.2023 11:26:19.141 [Debug] [PyboardSharpTransport] : received bytes: 0x45;0x53;0x50;0x2D;0x52;0x4F;0x4D ...
```
I know that the 0x0D in front of 0x03 is redundant. Didn't change anything having it there or not.

**Comment by TRadigk** [comment]
I thank everyone for participating 🙏. I got it.

## Resolution (and how I got there)
I finally started a series of tests varying many different serial port options (baudrate, stopbits, handshake, dtrenable, rtsenable) to find no combination that works with every MCU. But stubborn as I am, I took another look into pyboard transport, because it always showed to be working.
During my measurement more often the ESP32 with CH340 showed to be the odd one and thus I used this board to connect with pyboard, debugger attached. While stepping through I noticed that actually this part of code was addressed:
```Python
portinfo = list(serial.tools.list_ports.grep(device))  # type: ignore
if portinfo and portinfo[0].manufacturer != "Microsoft":
    # ESP8266/ESP32 boards use RTS/CTS for flashing and boot mode selection.
    # DTR False: to avoid using the reset button will hang the MCU in bootloader mode
    # RTS False: to prevent pulses on rts on serial.close() that would POWERON_RESET an ESPxx
    self.serial.dtr = False  # DTR False = gpio0 High = Normal boot
    self.serial.rts = False  # RTS False = EN High = MCU enabled
```
And since my C# driver never differentiated between the driver manufacturers I decided to implement this feature anyway. As the pyboard transport does I now switch off Dtr and Rts when the driver has not manufacturer "Microsoft", which is true for the _wch_ driver used for the ESP32 CH340 board. This is the C# code for everyone interested:

Note: in .net (included in .net framework) the additional nuget package System.Management is required.
```Csharp
private bool IsMicrosoftDriver(string comPort)
{
    var objSearcher = new ManagementObjectSearcher("root\\CIMV2",
                                                   "SELECT * FROM Win32_PnPEntity WHERE ClassGuid=\"{4d36e978-e325-11ce-bfc1-08002be10318}\"");

    var objCollection = objSearcher.Get();

    var providedComPorts = new Dictionary<string, string>();

    foreach (ManagementObject obj in objCollection)
    {
        string rawName = (string)obj["Name"];
        int nameStart = rawName.IndexOf('(') + 1;
        int nameStop = rawName.IndexOf(")", StringComparison.Ordinal);
        string comName = rawName.Substring(nameStart, nameStop - nameStart);
        providedComPorts.Add(comName, (string)obj["Manufacturer"]);
    }

    return providedComPorts[comPort] == "Microsoft";
}
```

**Reply by Josverl** [reply]
Happy to read that those changes and comments are of use to you
. Parts of them are in pymakr as well btw

If you want I have some automated tests for reset and reconnect cases as well. 

Note that recently there are new serial chips being used. Thus far the "non Microsoft" seems to hold true as intended, but frankly there is a bit of luck involved as well. 

Lastly, I suggest to raise this with Thonny so they can implement similar logic.

**Reply by Josverl** [reply]
Also, on the espxx series you can hook up a logic analiser to the rx/t pins to trace what is happening when the host is doing other things, and missing out on relevant output.

**Reply by TRadigk** [reply]
Actually I think there is no issue with Thonny (to raise). I think that my C# driver with the wrong connection properties (e.g. DtrEnable and / or DtsEnable with the MCU where it should be false) kind of blocked the serial port of windows.
Regarding the logic analyser: Yes you are right. Using the right tools to debug can be very helpful. But I never thought that I might be doing this kind of driver at home (I want to do other projects that are realized with micropython) and thus I didn't get one (yet). But the next time I will strongly consider getting one. In the end one must have to admit that it wouldn't have helped me anyway, since I have had no knowledge about the handling of dtr/rts in regards of ESP32. Adding to that both troublesome MCUs are ESP32 even with same micropython and only the serial port driver installed was different.
In order to make things easier I tested the standard windows driver for ESP32 with CH340 - doesn't work for even the simplest things.
But I am very grateful for your input, much appreciated 😊
